File: plateform/frontend/src/types/analytics.ts
export type AnalyticsRange = "24h" | "7d" | "30d";
export type AnalyticsInterval = "15m" | "30m" | "1h" | "6h" | "1d";
export type TrendDirection = "up" | "down" | "stable" | "unavailable";
export type MetricInterpretation = "higher_is_better" | "lower_is_better" | "neutral";

export interface AnalyticsFilters {
  range: AnalyticsRange;
  zone: string;
  sensor: string;
}

export interface AnalyticsFilterState {
  zone: string | null;
  sensor: string | null;
  interval: string | null;
}

export interface ComparedMetric {
  key: string;
  label: string;
  unit: string | null;
  current: number | null;
  previous: number | null;
  change: number | null;
  change_pct: number | null;
  trend: TrendDirection;
  interpretation: MetricInterpretation;
}

export interface NamedCount {
  key: string;
  label: string;
  count: number;
}

export interface TimeSeriesPoint {
  timestamp: string;
  pressure_bar: number | null;
  flow_m3h: number | null;
  packet_loss_pct: number | null;
  completeness_pct: number | null;
  reporting_sensors: number | null;
  reading_count: number | null;
  detections: number | null;
  incidents: number | null;
  mean_response_minutes: number | null;
}

export interface AnalyticsEnvelope {
  range: AnalyticsRange;
  start: string;
  end: string;
  generated_at: string;
  data_mode: "simulated";
  filters: AnalyticsFilterState;
}

export interface AnalyticsOverviewResponse extends AnalyticsEnvelope {
  kpis: ComparedMetric[];
  telemetry: Record<string, ComparedMetric>;
  events: Record<string, ComparedMetric>;
  operations: Record<string, ComparedMetric>;
  available_zones: string[];
  available_sensors: string[];
}

export interface AnalyticsTelemetryResponse extends AnalyticsEnvelope {
  interval: AnalyticsInterval;
  summary: Record<string, ComparedMetric>;
  series: TimeSeriesPoint[];
  reporting_sensors: number;
  stale_sensors: number;
  offline_sensors: number;
  selected_sensors: number;
}

export interface ZoneAnalyticsRow {
  zone: string;
  reporting_sensors: number;
  selected_sensors: number;
  telemetry_completeness_pct: number | null;
  average_pressure_bar: number | null;
  average_flow_m3h: number | null;
  detection_count: number;
  active_incident_count: number;
  average_asset_health: number | null;
  overdue_response_tasks: number;
  sufficient_data: boolean;
}

export interface AnalyticsZonesResponse extends AnalyticsEnvelope {
  items: ZoneAnalyticsRow[];
}

export interface AnalyticsDetectionsResponse extends AnalyticsEnvelope {
  interval: AnalyticsInterval;
  created: ComparedMetric;
  dismissed: ComparedMetric;
  promoted: ComparedMetric;
  promotion_rate: ComparedMetric;
  by_priority: NamedCount[];
  by_rule: NamedCount[];
  series: TimeSeriesPoint[];
}

export interface AnalyticsIncidentsResponse extends AnalyticsEnvelope {
  interval: AnalyticsInterval;
  created: ComparedMetric;
  resolved: ComparedMetric;
  false_alarms: ComparedMetric;
  by_severity: NamedCount[];
  by_classification: NamedCount[];
  by_status: NamedCount[];
  series: TimeSeriesPoint[];
}

export interface AnalyticsOperationsResponse extends AnalyticsEnvelope {
  interval: AnalyticsInterval;
  mean_acknowledgement_minutes: ComparedMetric;
  median_acknowledgement_minutes: ComparedMetric;
  mean_response_start_minutes: ComparedMetric;
  mean_resolution_minutes: ComparedMetric;
  open_response_tasks: ComparedMetric;
  completed_response_tasks: ComparedMetric;
  overdue_response_tasks: ComparedMetric;
  series: TimeSeriesPoint[];
}

export interface AssetHealthBucket {
  key: string;
  label: string;
  count: number;
}

export interface AnalyticsAssetsResponse extends AnalyticsEnvelope {
  health_bands: AssetHealthBucket[];
  operational_status: NamedCount[];
  connectivity: NamedCount[];
  average_health: number | null;
  sensor_count: number;
  total_assets: number;
}

export interface AnalyticsBundle {
  overview: AnalyticsOverviewResponse;
  telemetry: AnalyticsTelemetryResponse;
  zones: AnalyticsZonesResponse;
  detections: AnalyticsDetectionsResponse;
  incidents: AnalyticsIncidentsResponse;
  operations: AnalyticsOperationsResponse;
  assets: AnalyticsAssetsResponse;
}

File: plateform/backend/app/schemas/analytics.py
from datetime import datetime
from typing import Literal

from pydantic import BaseModel, Field

from app.services.analytics_constants import MetricInterpretation, TrendDirection


class AnalyticsFilters(BaseModel):
    zone: str | None = None
    sensor: str | None = None
    interval: str | None = None


class AnalyticsEnvelope(BaseModel):
    range: str
    start: datetime
    end: datetime
    generated_at: datetime
    data_mode: Literal["simulated"] = "simulated"
    filters: AnalyticsFilters


class ComparedMetric(BaseModel):
    key: str
    label: str
    unit: str | None = None
    current: float | None = None
    previous: float | None = None
    change: float | None = None
    change_pct: float | None = None
    trend: TrendDirection
    interpretation: MetricInterpretation = "neutral"


class NamedCount(BaseModel):
    key: str
    label: str
    count: int = Field(ge=0)


class TimeSeriesPoint(BaseModel):
    timestamp: datetime
    pressure_bar: float | None = None
    flow_m3h: float | None = None
    packet_loss_pct: float | None = None
    completeness_pct: float | None = None
    reporting_sensors: int | None = None
    reading_count: int | None = None
    detections: int | None = None
    incidents: int | None = None
    mean_response_minutes: float | None = None


class AnalyticsOverviewResponse(AnalyticsEnvelope):
    kpis: list[ComparedMetric]
    telemetry: dict[str, ComparedMetric]
    events: dict[str, ComparedMetric]
    operations: dict[str, ComparedMetric]
    available_zones: list[str]
    available_sensors: list[str]


class AnalyticsTelemetryResponse(AnalyticsEnvelope):
    interval: str
    summary: dict[str, ComparedMetric]
    series: list[TimeSeriesPoint]
    reporting_sensors: int
    stale_sensors: int
    offline_sensors: int
    selected_sensors: int


class ZoneAnalyticsRow(BaseModel):
    zone: str
    reporting_sensors: int
    selected_sensors: int
    telemetry_completeness_pct: float | None = None
    average_pressure_bar: float | None = None
    average_flow_m3h: float | None = None
    detection_count: int
    active_incident_count: int
    average_asset_health: float | None = None
    overdue_response_tasks: int
    sufficient_data: bool


class AnalyticsZonesResponse(AnalyticsEnvelope):
    items: list[ZoneAnalyticsRow]


class AnalyticsDetectionsResponse(AnalyticsEnvelope):
    interval: str
    created: ComparedMetric
    dismissed: ComparedMetric
    promoted: ComparedMetric
    promotion_rate: ComparedMetric
    by_priority: list[NamedCount]
    by_rule: list[NamedCount]
    series: list[TimeSeriesPoint]


class AnalyticsIncidentsResponse(AnalyticsEnvelope):
    interval: str
    created: ComparedMetric
    resolved: ComparedMetric
    false_alarms: ComparedMetric
    by_severity: list[NamedCount]
    by_classification: list[NamedCount]
    by_status: list[NamedCount]
    series: list[TimeSeriesPoint]


class AnalyticsOperationsResponse(AnalyticsEnvelope):
    interval: str
    mean_acknowledgement_minutes: ComparedMetric
    median_acknowledgement_minutes: ComparedMetric
    mean_response_start_minutes: ComparedMetric
    mean_resolution_minutes: ComparedMetric
    open_response_tasks: ComparedMetric
    completed_response_tasks: ComparedMetric
    overdue_response_tasks: ComparedMetric
    series: list[TimeSeriesPoint]


class AssetHealthBucket(BaseModel):
    key: str
    label: str
    count: int = Field(ge=0)


class AnalyticsAssetsResponse(AnalyticsEnvelope):
    health_bands: list[AssetHealthBucket]
    operational_status: list[NamedCount]
    connectivity: list[NamedCount]
    average_health: float | None = None
    sensor_count: int
    total_assets: int
